File: leet/0889-construct-binary-tree-from-preorder-and-postorder-traversal/construct-binary-tree-from-preorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def constructFromPrePost(self, pre: List[int], post: List[int]) -> TreeNode:
        if pre is None or len(pre) == 0 or post is None or len(post) == 0 or len(pre) != len(post):
            return None;
        
        if len(pre) == 1:
            return TreeNode(pre[0])
        
        root = TreeNode(pre[0])
        left = pre[1]
        right = post[-2]
        
        if left == right:
            root.left = self.constructFromPrePost(pre[1:], post[0:-1])
        else:
            rightIndexInPre = -1
            for i in range(len(pre)):
                if pre[i] == right:
                    rightIndexInPre = i
                    break
                    
            leftInfexInPost = -1
            for i in range(len(post)):
                if post[i] == left:
                    leftInfexInPost = i
                    break
            
            root.left = self.constructFromPrePost(pre[1:rightIndexInPre], post[0 : leftInfexInPost + 1])
            root.right = self.constructFromPrePost(pre[rightIndexInPre:], post[leftInfexInPost + 1 : -1])
            
        return root
                

File: leet/0889-construct-binary-tree-from-preorder-and-postorder-traversal/construct-binary-tree-from-preorder-and-postorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} pre
 * @param {number[]} post
 * @return {TreeNode}
 */
var constructFromPrePost = function(pre, post) {
    if(pre.length > 0 && post.length > 0){ 
        let ret = new TreeNode(pre[0])
        if(pre.length == 1) return ret

        if(pre[1] == post[post.length - 2]){
            ret.left = constructFromPrePost(pre.slice(1), post.slice(0, post.length - 1))
            return ret
        }
        else{
            let idx = -1
            for (let i = 0; i < post.length - 1; i++){
                if(pre[1] == post[i]) {
                    idx = i
                    break
                }
            }
            
            let rIdx = -1
            for (let i = 1; i < pre.length; i++){
                if(post[post.length - 2] == pre[i]){
                    rIdx = i
                    break
                }
            }
            
            if(idx > -1)
                ret.left = constructFromPrePost(pre.slice(1, rIdx + 1), post.slice(0, idx + 1))
            if(rIdx > -1)
                ret.right = constructFromPrePost(pre.slice(rIdx), post.slice(idx, -1))
            
            return ret
        }
    }
};
